sdk/metric/meter.go

```diff
@@ -93,28 +104,28 @@ func (r *meterRegistry) Range(f func(*meter) bool) {
 type meter struct {
 	instrumentation.Scope
 
-	// TODO (#2815, 2814): implement.
+	intRegistry   *pipelineRegistry[int64]
+	floatRegistry *pipelineRegistry[float64]
 }
 
 // Compile-time check meter implements metric.Meter.
 var _ metric.Meter = (*meter)(nil)
 
 // AsyncInt64 returns the asynchronous integer instrument provider.
 func (m *meter) AsyncInt64() asyncint64.InstrumentProvider {
-	// TODO (#2815): implement.
-	return nil
+	return asyncInt64Provider{scope: m.Scope, registry: m.intRegistry}
 }
 
 // AsyncFloat64 returns the asynchronous floating-point instrument provider.
 func (m *meter) AsyncFloat64() asyncfloat64.InstrumentProvider {
-	// TODO (#2815): implement.
-	return nil
+	return asyncFloat64Provider{scope: m.Scope, registry: m.floatRegistry}
 }
 
 // RegisterCallback registers the function f to be called when any of the
 // insts Collect method is called.
 func (m *meter) RegisterCallback(insts []instrument.Asynchronous, f func(context.Context)) error {
-	// TODO (#2815): implement.
+	// Because the pipelines are shared only one of the registries needs to be invoked
+	m.intRegistry.registerCallback(f)
```

Comment:
This has a bad smell, our pipeline registry does not match our needs here.

It seems like the `pipelineRegistry` type does not need to be defined generic types. It is a requirement that is coming from the `createAggregators` method. However, there is already also [`createAggregators`](https://github.com/open-telemetry/opentelemetry-go/blob/8580a3525ccc01ee141264a910d19cf1d7ec4f8a/sdk/metric/pipeline.go#L243) function that is defined with generic value types. If we changed that function to accept a `pipelineRegistry` instead of a `Reader` the method could be dropped, the `pipelineRegistry` unified, and instruments would just call the function directly.

I'm fine addressing this in a separate PR instead of here.

Reply:
This should be addressed in a separate PR.

Reply:
You are right, this is a smell, and I think the solution is not to have the Registry be generic.

My solution for this I think will take the current `createAggregators` and rename it to `createAggregatorsForViews`, which is what it really does.  Then have two methods on the Registry, `CreateIntAggregators` and `CreateFloatAggregators` which calls a new `createAggreagtors` which has the logic that is the current `pipelineRegistry.createAggregators`

Reply:
https://github.com/open-telemetry/opentelemetry-go/issues/3112